src/kvstore/comm.h

```diff
@@ -635,6 +656,302 @@ class CommDevice : public Comm {
+                int priority) override {
+    std::vector<Engine::VarHandle> const_vars;
+    std::vector<Engine::VarHandle> mutate_vars;
+    for (size_t i = 0; i < dst.size(); ++i) {
+        mutate_vars.push_back(dst[i]->var());
+    }
+    Engine::Get()->PushSync([this](RunContext rctx) {
+          for (auto cur : nccl_data_) {
+            CUDA_CALL(cudaSetDevice(cur.second.dev_id));
+            CUDA_CALL(cudaStreamSynchronize(cur.second.stream));
+          }
+        },
+        Context::CPU(),
+        const_vars,
+        mutate_vars,
+        FnProperty::kCPUPrioritized,
+        priority,
+        PROFILER_MESSAGE("KVStoreStreamSync"));
+  }
+
+  void CommSync(const std::vector<NDArray>& dst,
+                int priority) override {
+    std::vector<Engine::VarHandle> const_vars;
+    std::vector<Engine::VarHandle> mutate_vars;
+    for (size_t i = 0; i < dst.size(); ++i) {
+        mutate_vars.push_back(dst[i].var());
+    }
+    Engine::Get()->PushSync([this](RunContext rctx) {
+          for (auto cur : nccl_data_) {
+            CUDA_CALL(cudaSetDevice(cur.second.dev_id));
+            CUDA_CALL(cudaStreamSynchronize(cur.second.stream));
+          }
+        },
+        Context::CPU(),
+        const_vars,
+        mutate_vars,
+        FnProperty::kCPUPrioritized,
+        priority,
+        PROFILER_MESSAGE("KVStoreStreamSync"));
+  }
+
+  void BroadcastRowSparse(int key, const NDArray& src,
+                          const std::vector<std::pair<NDArray*, NDArray>>& dst,
+                          const bool use_copy,
+                          const int priority) override {
+    LOG(FATAL) << "NCCL kvstore does not support sparse storage type";
+  }
+
+  void Broadcast(int key, const NDArray& src,
+                 const std::vector<NDArray> dst, int priority) override {
+    if (!inited_) {
+      // copy to a random device first
+      int dev_id = key % dst.size();
+      CopyFromTo(src, dst[dev_id], priority);
+      for (size_t i = 0; i < dst.size(); ++i) {
+        if (i != static_cast<size_t>(dev_id)) {
+          CopyFromTo(dst[dev_id], dst[i], priority);
+        }
+      }
+    } else {
+      auto& buf = merge_buf_[key];
+      int root = src.ctx().dev_id;
+      assert(root == buf.ctx().dev_id);
+      size_t root_id = -1;
+      for (size_t i = 0; i < dst.size(); ++i) {
+        if (dst[i].ctx().dev_id == root) {
+          root_id = i;
+          break;
+        }
+      }
+      std::vector<int> dev_ids;
+      for (size_t i = 0; i < dst.size(); ++i) {
+        auto& bcast = (i == root_id) ? src : dst[i];
+        dev_ids.push_back(bcast.ctx().dev_id);
+      }
+      std::sort(dev_ids.begin(), dev_ids.end());
+      CHECK(device_ids_ == dev_ids) << "NCCL KVStore supports only single set of devices";
+      CopyFromTo(src, dst[root_id], priority);
+      if (dst.size() == 1) return;
+      std::vector<Engine::VarHandle> mutable_vars;
+      for (size_t i = 0; i < dst.size(); ++i) {
+        if ( i != root_id)
```

Comment:
Please add braces around line 848. This style has potential for future bugs :)

Reply:
Ok